SATSubsumption/SATSubsumptionAndResolution.hpp

```diff
@@ -0,0 +1,549 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file SATSubsumptionAndResolution.cpp
+ * Defines class SATSubsumptionAndResolution.
+ */
+#ifndef SAT_SUBSUMPTION_RESOLUTION_HPP
+#define SAT_SUBSUMPTION_RESOLUTION_HPP
+
+#include "Kernel/Clause.hpp"
+#include "Lib/STL.hpp"
+
+#include "./subsat/subsat.hpp"
+
+/// @brief If 1, prints the SAT clauses added to the solver on the standard output
+#define PRINT_CLAUSES_SUBS 0
+/// @brief If 1, prints some comments about the subsumption resolution process
+#define PRINT_CLAUSE_COMMENTS_SUBS 0
+
+namespace SATSubsumption {
+
+/**
+ * Class implementing the simplifying rules of subsumption and subsumption resolution using a SAT solver.
+ *
+ * The subsumption and subsumption resolution are described in the papers:
+ * - 2022: "First-Order Subsumption via SAT Solving." by Jakob Rath, Armin Biere and Laura Kovács
+ * - 2023: "SAT-Based Subsumption Resolution" by Robin Coutelier, Jakob Rath, Michael Rawson and Laura Kovács
+ * - 2024: "SAT Solving for Variants of First-Order Subsumption" by Robin Coutelier, Jakob Rath, Michael Rawson, Armin Biere and Laura Kovács
+ */
+class SATSubsumptionAndResolution {
+#if VDEBUG
+  // Make it public to allow unit testing
```

Comment:
I didn't touch the unit testing so I'm not sure, but it doesn't seem that the tests are in a class. We probably would have to add a wrapper class to the unit test file for this. @RobCoutel what do you think?

Reply:
I think I saw some other classes manage it with `friend`, which is where I got the idea from, but could be wrong. No big deal either way.

Reply:
It is maybe a bit dirty. The reason I did this is that the unit tests are not really a class (with the current framework). So to access private methods, this is the easiest. I am not sure if this is a bit problem though.

You could do this with a `friend`, but this would require to have a test class. Which is not the way it is currently done.

Reply:
I maybe hallucinated the `friend` test thing, I can't find it now.